### PR TITLE
CPP: Add query for CWE-1126: Declaration of Variable with Unnecessarily Wide Scope

### PR DESCRIPTION
**Pull request by ihsinme**
Good day.
in this request I am looking at the error of using a new variable inside a loop.
in a situation where a variable with the same name participates in a loop condition.

the most dangerous situation is the while loop in the absence of changes to the variables participating in the condition.
At this stage, I tried to minimize false positives as much as possible.
I left the situation of changing a variable through a function call. as I considered it a rather rare practice. if you disagree with me, I will listen to you

current PR.
sleuthkit/sleuthkit#2329

### CODE REVIEW

File: cpp/ql/src/experimental/Security/CWE/CWE-1126/DeclarationOfVariableWithUnnecessarilyWideScope.c [hunk 1]
```diff
@@ -0,0 +1,14 @@
+while(intIndex > 2)
+{
+  ...
+  intIndex--;
+  ...
+} // GOOD: coreten cycle
```
Comment (geoffw0) [inline cpp/ql/src/experimental/Security/CWE/CWE-1126/DeclarationOfVariableWithUnnecessarilyWideScope.c:6]:
```suggestion
} // GOOD: correct cycle
```
?

File: cpp/ql/src/experimental/Security/CWE/CWE-1126/DeclarationOfVariableWithUnnecessarilyWideScope.qhelp [hunk 2]
```diff
@@ -0,0 +1,26 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+<overview>
+<p>Using variables with the same name is dangerous. However, such a situation inside the while loop can lead to a violation of the accessibility of the program. Requires the attention of developers.</p>
```
Comment (geoffw0) [inline cpp/ql/src/experimental/Security/CWE/CWE-1126/DeclarationOfVariableWithUnnecessarilyWideScope.qhelp:6]: > However, such a situation inside the while loop can lead to a violation of the accessibility of the program.

Please could you rephrase this sentence, I'm not sure what you mean by "the accessibility of the program".

There's a similar problem in the query `@description` as well.

File: cpp/ql/test/experimental/query-tests/Security/CWE/CWE-1126/semmle/tests/test.c [hunk 3]
```diff
@@ -0,0 +1,15 @@
+void workFunction_0(char *s) {
+  int intIndex = 10;
+  char buf[80];
+  while(intIndex > 2) // GOOD
+  {
+    buf[intIndex] = 1;
+    intIndex--;
+  }
+  while(intIndex > 2)
```
Comment (geoffw0) [inline cpp/ql/test/experimental/query-tests/Security/CWE/CWE-1126/semmle/tests/test.c:9]: Could we use a distinct variable for the second loop?  As it is, `intIndex` is already `2` following execution of the first loop so the second loop would not be entered.

File: cpp/ql/src/experimental/Security/CWE/CWE-1126/DeclarationOfVariableWithUnnecessarilyWideScope.ql [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+ * @description Using variables with the same name is dangerous.
+ *              However, such a situation inside the while loop can lead to a violation of the accessibility of the program.
+ *              Requires the attention of developers.
+ * @kind problem
+ * @id cpp/errors-when-using-variable-declaration-inside-loop
+ * @problem.severity warning
+ * @precision medium
+ * @tags correctness
+ *       security
+ *       external/cwe/cwe-1126
```
Comment (geoffw0) [inline cpp/ql/src/experimental/Security/CWE/CWE-1126/DeclarationOfVariableWithUnnecessarilyWideScope.ql:12]: I think this is more an issue of correctness than security, and I'm not convinced the CWE-1126 tag is accurate.

File: cpp/ql/src/experimental/Security/CWE/CWE-1126/DeclarationOfVariableWithUnnecessarilyWideScope.ql [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+ */
+
+import cpp
+
+/**
+ * Errors when using a variable declaration inside a loop.
+ */
+class DangerousWhileLoop extends WhileStmt {
```
Comment (geoffw0) [inline cpp/ql/src/experimental/Security/CWE/CWE-1126/DeclarationOfVariableWithUnnecessarilyWideScope.ql:20]: I think you could include `DoStmt` as well.

File: cpp/ql/src/experimental/Security/CWE/CWE-1126/DeclarationOfVariableWithUnnecessarilyWideScope.ql [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+  Expr exp;
+  Declaration dl;
+
+  DangerousWhileLoop() {
+    this = dl.getParentScope().(BlockStmt).getParent*() and
+    exp = this.getCondition().getAChild*() and
+    not exp instanceof PointerFieldAccess and
+    not exp instanceof ValueFieldAccess and
+    exp.toString() = dl.getName() and
```
Comment (geoffw0) [inline cpp/ql/src/experimental/Security/CWE/CWE-1126/DeclarationOfVariableWithUnnecessarilyWideScope.ql:29]: Please don't use `toString` in query logic.  The behaviour of `toString` can be unpredictable and may change in future without consideration for uses like this.  You _can_ use `getName` to compare the names of things (variables, functions etc) as `getName` will always return the name.

Are we looking for a `VariableAccess va` with `va.getTarget().getName() = dl.getName()`?

Can we then get rid of some of the exceptions (lines beginning with `not`)?

File: cpp/ql/src/experimental/Security/CWE/CWE-1126/DeclarationOfVariableWithUnnecessarilyWideScope.ql [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    not exp.getParent*() instanceof CrementOperation and
+    not exp.getParent*() instanceof Assignment and
+    not exp.getParent*() instanceof FunctionCall
+  }
+
+  Declaration getDeclaration() { result = dl }
+
+  /** Holds when there are changes to the variables involved in the condition. */
```
Comment (geoffw0) [inline cpp/ql/src/experimental/Security/CWE/CWE-1126/DeclarationOfVariableWithUnnecessarilyWideScope.ql:37]: Could we have an example in the tests where this exception matters?

**Review by geoffw0 — COMMENTED** [review]
This reminds me of the queries in `Best Practices/Hiding`, especially `cpp/declaration-hides-variable`.  Your query appears to be a more specialized version looking for instances where there's evidence the hiding is causing a real problem.

File: cpp/ql/src/experimental/Security/CWE/CWE-1126/DeclarationOfVariableWithUnnecessarilyWideScope.qhelp (hunk 2 above)
Comment (ihsinme) [reply cpp/ql/src/experimental/Security/CWE/CWE-1126/DeclarationOfVariableWithUnnecessarilyWideScope.qhelp:6]: sorry for these mistakes.
I am working on my English, but I am ashamed in front of you every time.

I meant the impact of this error on the creation of an infinite loop exhausting resources.
is it sufficient to replace `accessibility` with` availability`?

File: cpp/ql/test/experimental/query-tests/Security/CWE/CWE-1126/semmle/tests/test.c (hunk 3 above)
Comment (ihsinme) [reply cpp/ql/test/experimental/query-tests/Security/CWE/CWE-1126/semmle/tests/test.c:9]: thank you for noticing this error.
but considering your request to increase the number of test cases, I propose to leave the variable in a single form.

File: cpp/ql/src/experimental/Security/CWE/CWE-1126/DeclarationOfVariableWithUnnecessarilyWideScope.ql (hunk 4 above)
Comment (ihsinme) [reply cpp/ql/src/experimental/Security/CWE/CWE-1126/DeclarationOfVariableWithUnnecessarilyWideScope.ql:12]: The main influence vector I see is replacing the scope of a variable in the loop condition with the scope of a variable declared inside the loop.
this brings me to the real issue of security, resource depletion and `CWE-1126`.

File: cpp/ql/src/experimental/Security/CWE/CWE-1126/DeclarationOfVariableWithUnnecessarilyWideScope.ql (hunk 5 above)
Comment (ihsinme) [reply cpp/ql/src/experimental/Security/CWE/CWE-1126/DeclarationOfVariableWithUnnecessarilyWideScope.ql:20]: when designing this query, I evaluated this behavior for all kinds of loops.
but `for` loops are protected by their structure from this problem.
`dowhile` loops are not subject to overriding the scope that the query is looking for.
so I have to leave only the `while`.
expansion is possible, but it will reduce the detection rate of real problems.
although I agree that overall stylistic detection will be improved.

File: cpp/ql/src/experimental/Security/CWE/CWE-1126/DeclarationOfVariableWithUnnecessarilyWideScope.ql (hunk 6 above)
Comment (ihsinme) [reply cpp/ql/src/experimental/Security/CWE/CWE-1126/DeclarationOfVariableWithUnnecessarilyWideScope.ql:29]: thanks for that suggestion.
I used it partially.
unfortunately I could not remove additional conditions.

**Comment by ihsinme** [comment]
thanks for your comments.
they are very helpful.

File: cpp/ql/src/experimental/Security/CWE/CWE-1126/DeclarationOfVariableWithUnnecessarilyWideScope.qhelp (hunk 2 above)
Comment (geoffw0) [reply cpp/ql/src/experimental/Security/CWE/CWE-1126/DeclarationOfVariableWithUnnecessarilyWideScope.qhelp:6]: I think you described it clearly in your reply (above), but neither word (`accessibility` or `availability`) conveys this meaning.  Perhaps:
```
However, such a situation inside the while loop can create an infinite loop exhausting resources.
```

File: cpp/ql/test/experimental/query-tests/Security/CWE/CWE-1126/semmle/tests/test.c (hunk 3 above)
Comment (geoffw0) [reply cpp/ql/test/experimental/query-tests/Security/CWE/CWE-1126/semmle/tests/test.c:10]: I'm happy with your fix.  :+1:

File: cpp/ql/src/experimental/Security/CWE/CWE-1126/DeclarationOfVariableWithUnnecessarilyWideScope.ql [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+  predicate isUseThisVariable() {
+    exists(Variable v |
+      this.getCondition().getAChild*().(VariableAccess).getTarget() = v and
```
Comment (geoffw0) [inline cpp/ql/src/experimental/Security/CWE/CWE-1126/DeclarationOfVariableWithUnnecessarilyWideScope.ql:40]: Can `this.getCondition().getAChild*()` be replaced with `exp`?

File: cpp/ql/src/experimental/Security/CWE/CWE-1126/DeclarationOfVariableWithUnnecessarilyWideScope.qhelp (hunk 2 above)
Comment (ihsinme) [reply cpp/ql/src/experimental/Security/CWE/CWE-1126/DeclarationOfVariableWithUnnecessarilyWideScope.qhelp]: thank you

File: cpp/ql/src/experimental/Security/CWE/CWE-1126/DeclarationOfVariableWithUnnecessarilyWideScope.ql (hunk 8 above)
Comment (ihsinme) [reply cpp/ql/src/experimental/Security/CWE/CWE-1126/DeclarationOfVariableWithUnnecessarilyWideScope.ql]: sure

**Comment by geoffw0** [comment]
Started a test run on LGTM: https://lgtm.com/query/3978555328191505856/

**Comment by ihsinme** [comment]
> Started a test run on LGTM: https://lgtm.com/query/3978555328191505856/

good day.
I looked at the launch results.
I would like to point out the following
1.In the project `ArtifexSoftware/ghostpdl`, the detection on the change of a variable did not work, I need some time to figure it out.
2.` libretro/libretro-uae` `WohlSoft/PGE-Project`` php/php-src` projects showed false detection as I hastily replaced `this.getCondition().GetAChild*()`. this eliminated the control of changes for all variables in the loop condition. I will probably return the old version.

File: cpp/ql/src/experimental/Security/CWE/CWE-1126/DeclarationOfVariableWithUnnecessarilyWideScope.ql [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+          aexp = this.getStmt().getAChild*() and
+          (
+            aexp.getLValue().(ArrayExpr).getArrayBase().(VariableAccess).getTarget() = v
+            or
+            aexp.getLValue().(VariableAccess).getTarget() = v
+          )
+        )
+        or
+        exists(CrementOperation crm |
+          crm = this.getStmt().getAChild*() and
```
Comment (geoffw0) [inline cpp/ql/src/experimental/Security/CWE/CWE-1126/DeclarationOfVariableWithUnnecessarilyWideScope.ql:52]: I think some of your false positives are being caused by this line.  `Stmt.getAChild()` returns an `Element` not a `Stmt`, so it doesn't recurse as you might expect.

Try:
```
this = crm.getEnclosingStmt().getParentStmt*()
```

File: cpp/ql/src/experimental/Security/CWE/CWE-1126/DeclarationOfVariableWithUnnecessarilyWideScope.ql [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+          crm.getOperand().(VariableAccess).getTarget() = v
+        )
+      )
+    )
+  }
+}
+
+from DangerousWhileLoop lp
+where not lp.isUseThisVariable()
+select lp.getDeclaration(), "A variable with this name is used in the loop condition."
```
Comment (geoffw0) [inline cpp/ql/src/experimental/Security/CWE/CWE-1126/DeclarationOfVariableWithUnnecessarilyWideScope.ql:62]: It might help to debug problems if you have the query message link to the loop it's talking about:
```suggestion
select lp.getDeclaration(), "A variable with this name is used in the $@ condition.", lp, "loop"
```

**Comment by ihsinme** [comment]
thanks for the note with recursion.
with its help, I managed to get rid of a number of logical negatives in the class constructor.
I also added two new loops to the test file.
I would pay attention to an interesting discovery where there is clearly a confusion with the logic of the loop
https://lgtm.com/projects/g/RediSearch/RediSearch/snapshot/0e0a13957cf14e368049dda4a01b33b76d40bd49/files/src/rmutil/util.c?sort=name&dir=ASC&mode=heatmap#xfa6eebb857c378bb:1

**Comment by ihsinme** [comment]
I do not really understand what you mean.
if this is a reaction to the analysis of the detection, sorry, I deleted it.
could you please delete your comment

**Comment by geoffw0** [comment]
Changes look good to me, we seem to have lost some false positive results including https://lgtm.com/projects/g/php/php-src/snapshot/1d919127b62b1c3d0c7e50b7925896dfa7831d0f/files/ext/opcache/Optimizer/zend_optimizer.c?sort=name&dir=ASC&mode=heatmap#L623 and https://lgtm.com/projects/g/WohlSoft/PGE-Project/snapshot/4443036d0b7cf288c4d962b1faf61f053d07d279/files/Engine/common_features/QTranslatorX/qm_translator.cpp?sort=name&dir=ASC&mode=heatmap#L1040

**Review by geoffw0 — APPROVED** [review]
I think you've addressed all of the suggestions, the QL is satisfactory and the tests are excellent.

Results on LGTM: https://lgtm.com/query/8106651676569760712/ (very few, probably because we're looking for cases that are definitely causing problems rather than just any instances of reusing variable names).
